keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1587 @@
+optional properties; policy definition authors should only include the parts of
+the "match schema template" that are useful for configuring a particular policy.
+
+(Also, by allowing the "matching schema template" in configuration CRDs to be a
+omit optional properties, this API is future proofed against the addition of
+other match related properties in the future).
+
+"matching schema template":
+
+```go
+// TODO: Add this as a struct into the Kubernetes codebase so it can easily be
+// imported?
+type Match struct {
+  Rules []admissionregisterationv1.RuleWithOperations `json:"rules,omitempty" protobuf:"bytes,1,rep,name=rules"`
+  NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty" protobuf:"bytes,2,opt,name=namespaceSelector"`
+  ObjectSelector *metav1.LabelSelector `json:"objectSelector,omitempty" protobuf:"bytes,3,opt,name=objectSelector"`
+}
+```
+
+Example usage:
+
+```go
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type ReplicaLimit struct {
+  MaxReplicas int32 `json:"maxReplicas" protobuf:"varint,1,name=maxReplicas"`
+
+  Match Match `json:"match,omitempty" protobuf:"bytes,2,name=match"`
+}
+```
+
+For comparison (provided by @liggitt):
+
+- RBAC, priority & fairness, and admission webhook configurations all use
+ordered lists of rules, and first match allows/matches.
+- Audit configuration also uses an ordered list of rules, but can make a match
+exclude auditing... this approach could address the "ability to skip specific
+resource types" mentioned above.
+
+TODO: What exact matching mechanisms will be supported? Right now this KEP
+shows the use of the same match rules that admission webhooks use.
+
+TODO: Should policy definitions also have match criteria? Match criteria at the
+policy definition level could be used constrain what resources the policy may
+apply to across the entire cluster (like the above "kind constraint" but more
+generally). This would allow policy authors to constrain use of the policy to
+the resources it was designed for.
+
+TODO: Should "standalone policies" be supported? I.e. should it be possible to
+declare a policy definition that has no `spec.config` but has it's own match
+criteria? Right now a policy requires a minimum of three resources to do
+anything (policy definition, configuration CRD, at least one policy
+configuration).
+
+`MatchPolicy` will implicitly be `Equivalent`, and will not be configurable,
+but it will be possible to opt-out of the behavior by inspecting the
+GVK of the admitted resource in the CEL expression.
+
+xref:
+https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-rules
+
+#### Type safety
+
+To keep failure policy easy to reason about, and to continue to use CEL in a
+type-safe way we propose:
+
+- If a ValidatingAdmissionPolicy has a `spec.constraints` that matches a single
```

Comment:
This seems like a pain for "pod-specy-things"

Reply:
epod-specy things are a tricky case. Alternatives that I've thought about:

1. Use a single configuration CRD, define multiple ValidatingAdmissionPolicy, one for each GVK containing the pod spec to be validated. Typesafe but requires lots of policy definitions.
2. Write a single CEL expression that can dynamically validates all the GVKs that contain a pod spec (CEL has no switch statement.. this gets ugly quick)
3. Introduce a capability to run CEL expressions on types nested within resources being admitted (e.g. `{scope: io.k8s.api.core.v1.PodTemplateSpec, expression: '...'`)

(1) can be made a bit more egronomic by grouping cel expressions by the type they validated so that a single policy could have validation expressions for multiple GVKs

Reply:
What about just saying, if it matches a single type then you get type safety, and if it can handle multiple types you don't get type checking. Is that too much of a compromise?

Reply:
> if it matches a single type then you get type safety, and if it can handle multiple types you don't get type checking

We might end up here. I took a more careful look at all the requirements that impact type safety and I'm now starting to take the opinion that type checking should be informational-- we should emit warnings only.  See the updated "Type Checking" section.

Reply:
WRT pod specy things... if the goal is to enforce policy, should we only enforce it against the pod itself?

Another issue with pod-specy things is that there is no guarantee that metadata will be identical between, say, the deployment and the pod it creates, and there is the possibility that a mutating webhook would bring a pod into compliance that doesn't touch deployments.

All of this means that trying to guess whether a higher-order resource would create a valid lower-order resource is brittle at best. This is why things like PSA don't strictly enforce policies against higher-order resources.

If we do want to enforce policies against a higher-order resource, a better approach may be to create a mock of the resource it is presumed to create, so we can have mutation webhooks, etc. apply.

I'm definitely not against multi-GVK policies, but it does present a scaling issue. Baking the operator behavior into each policy definition would make it expensive to accommodate changes to operator behavior.

Reply:
I've simplified this but retained exclude support.